File: Advanced Programming Methods/Toy Language C#/A2/UI/TextMenu.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace A2.UI
{
    //hashmap that keeps value-command pairs
    public class TextMenu
    {
        private Dictionary<string, Command> commands;

        public TextMenu()
        {
            commands = new Dictionary<string, Command>();
        }

        public void addCommand(Command c)
        {
            commands.Add(c.getKey(), c);
        }

        private void printMenu()
        {
            foreach (Command com in this.commands.Values)
            {
                Console.WriteLine($"{com.getKey()}:{com.getDescription()}");
            }
        }

        public void show()
        {
            while (true)
            {
                printMenu();
                Console.Write("Input the option: ");
                String key = Console.ReadLine();
                Command com;
                commands.TryGetValue(key, out com);
                if (com == null)
                {
                    Console.WriteLine("Invalid Option");
                    continue;
                }
                com.execute();
            }
        }
    }
}

File: Advanced Programming Methods/Toy Language Java/src/UI/Menu/TextMenu.java
package UI.Menu;

import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

//hashmap that keeps value-command pairs
public class TextMenu {
    private Map<String, Command> commands;

    public TextMenu(){
        commands = new HashMap<>();
    }

    public void addCommand(Command c){
        commands.put(c.getKey(),c);
    }

    private void printMenu() {
        for (Command com : this.commands.values()) {
            String line = String.format("%4s:%s",com.getKey(),com.getDescription());
            System.out.println(line);
        }
    }

    public void show(){
        Scanner scanner = new Scanner(System.in);
        while(true){
            printMenu();
            System.out.print("Input the option: ");
            String key = scanner.nextLine();
            Command com = commands.get(key);
            if(com == null){
                System.out.println("Invalid Option");
                continue;
            }
            com.execute();
        }
    }
}